ctapipe/reco/sklearn.py

```diff
@@ -507,6 +501,325 @@ def predict_table(self, key, table: Table) -> Table:
+            self.subarray = subarray
+
+    def _new_models(self):
+        norm_regressor = SUPPORTED_REGRESSORS[self.norm_cls](**self.norm_config)
+        sign_classifier = SUPPORTED_CLASSIFIERS[self.sign_cls](**self.sign_config)
+        return norm_regressor, sign_classifier
+
+    def _table_to_y(self, table, mask=None):
+        """
+        Extract target values as numpy array from input table
+        """
+        # make sure we use the unit that was used during training
+        if self.unit is not None:
+            norm_y = np.abs(table[mask][self.target].quantity.to_value(self.unit))
+        else:
+            norm_y = np.abs(np.array(table[self.target][mask]))
+
+        if self.log_target:
+            if np.any(norm_y <= 0):
+                raise ValueError("norm_y contains negative values, cannot apply log")
+            norm_y = np.log(norm_y)
+
+        sign_y = np.sign(np.array(table[self.target][mask]))
+        return norm_y, sign_y
+
+    def fit(self, key, table):
+        """
+        Create and fit new models for ``key`` using the data in ``table``.
+        """
+        self._norm_models[key], self._sign_models[key] = self._new_models()
+
+        X, valid = table_to_X(table, self.features, self.log)
+        self.unit = table[self.target].unit
+        norm_y, sign_y = self._table_to_y(table, mask=valid)
+        self._norm_models[key].fit(X, norm_y)
+        self._sign_models[key].fit(X, sign_y)
+
+    def write(self, path, overwrite=False):
+        path = pathlib.Path(path)
+
+        if path.exists() and not overwrite:
+            raise IOError(f"Path {path} exists and overwrite=False")
+
+        with path.open("wb") as f:
+            Provenance().add_output_file(path, role="ml-models")
+            joblib.dump(self, f, compress=True)
+
+    @classmethod
+    def read(cls, path, **kwargs):
+        with open(path, "rb") as f:
+            instance = joblib.load(f)
+
+        for attr, value in kwargs.items():
+            setattr(instance, attr, value)
+
+        if not isinstance(instance, cls):
+            raise TypeError(
+                f"{path} did not contain an instance of {cls}, got {instance}"
+            )
+
+        Provenance().add_input_file(path, role="ml-models")
+        return instance
+
+    @lazyproperty
+    def instrument_table(self):
+        return self.subarray.to_table("joined")
+
+    def _predict(self, key, table):
+        if key not in self._norm_models:
+            raise KeyError(
+                f"No regressor available for key {key},"
+                f" available regressors: {self._norm_models.keys()}"
+            )
+        if key not in self._sign_models:
+            raise KeyError(
+                f"No classifier available for key {key},"
+                f" available classifiers: {self._sign_models.keys()}"
+            )
+        X, valid = table_to_X(table, self.features, self.log)
+        norm_prediction = np.full(len(table), np.nan)
```

Comment:
you can use a single predictions array now and do `prediction[valid] *= self._sign_models[key].predict(X)` which will be faster and save memory.

Reply:
Done